llvm/docs/SYCLOffloadingDesign.rst

```diff
@@ -0,0 +1,677 @@
+FPGAs, and other processors.
+
+A SYCL implementation refers to a software version or revision of the SYCL
+programming model. It consists of the necessary tools, libraries, and runtime
+support that enable developers to write SYCL code and execute it on compatible
+hardware platforms. These are some of the popular SYCL implementations that are
+available. This document focuses on DPCPP by Intel which is a part of the oneAPI
+solution that supports Intel CPUs, GPUs, and FPGAs as well as NVIDIA and AMD
+GPUs. For more details, please look at `sycl-overview <https://www.khronos.org/sycl/>`_.
+
+  .. code-block:: c++
+    :linenos:
+    
+      #include <sycl/sycl.hpp>
+      using namespace sycl;
+      int main() {
+          queue Q;                        // The queue, Q, is the object that
+                                          // submits the task to a device.
+          int const size = 10;
+          buffer<int> A{ size };          // The buffer, A, is the memory used to
+                                          // transfer data between host and device.
+          Q.submit([&](handler& h) {      // The handler, h, is the object that contains
+                                          // the single_task function to be used.
+              accessor A_acc(A, h);       // The accessor, A_acc, is the object that
+                                          // efficiently accesses the buffer elements.
+              h.single_task([=]() {
+                  A_acc[5] = 77;
+                  });
+              });
+          host_accessor result(A);        // host_accessor is the object that allows
+                                          // the host to access the buffer memory.
+          for (int i = 0; i < size; i++)  // Print output
+              std::cout << result[i] << " "; std::cout << "\n";
+          return 0;
+      }
+
+Here, we present a code with a simple SYCL algorithm called single_task, which
+is equivalent in standard C++ would be std::thread, to gain a fundamental
+understanding of SYCL components. The construct single_task is used to define a
+unit of work that should be executed on a single processing element, typically a
+single CPU core or GPU thread. A SYCL kernel function is a lambda function that
+is passed as an argument to the single_task construct.
+This document describes the clang driver and code generation steps for creating
+offloading applications. Clang supports offloading to various architectures
+using programming models like CUDA, HIP, and OpenMP. Currently, Clang support
+for SYCL offloading is being added.
+
+Clang will support SYCL target offloading to several different architectures
+such as Intel CPUs and GPUs, NVPTX, and AMDGPU. Offloading code is generated by
+clang and then executed using SYCL runtime. This section describes the steps
+necessary to create a functioning device image that can be loaded by the SYCL
+runtime. {{ SYCL runtime reference? }}.
```

Comment:
It would be helpful to describe the relationship to OpenCL, CUDA, HIP, SPIR-V, etc... somewhere in this section.